core/src/ipld.rs

```diff
@@ -159,6 +161,104 @@ impl Ipld {
+            4 => Ipld::String(String::arbitrary(g)),
+            5 => Ipld::Bytes(Vec::arbitrary(g)),
+            6 => Ipld::List(Self::arbitrary_ipld_vec(g, &mut size)),
+            7 => Ipld::Map(Self::arbitrary_ipld_map(g, &mut size)),
+            8 => {
+                let cid = Cid::new_v1(
+                    u64::arbitrary(g),
+                    cid::multihash::Multihash::wrap(u64::arbitrary(g), &[u8::arbitrary(g)])
+                        .unwrap(),
+                );
+                Ipld::Link(cid)
+            }
+            // unreachable due to the fact that
+            // we know that the index is always < 9
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[cfg(feature = "arb")]
+impl Ipld {
+    /// Special version on `arbitrary` to battle possible recursion
+    pub fn arbitrary_ipld(g: &mut quickcheck::Gen, size: &mut usize) -> Self {
+        if *size == 0 {
+            return Ipld::Null;
+        }
+        *size -= 1;
+        let choose_vec = Vec::from_iter(0..9);
+        let index = g.choose(choose_vec.as_slice()).unwrap();
+        match index {
+            0 => Ipld::Null,
+            1 => Ipld::Bool(bool::arbitrary(g)),
+            2 => Ipld::Integer(i128::arbitrary(g)),
+            // Filter out problematic NaN values.
+            3 => {
+                let float_item = f64::arbitrary(g);
+                Ipld::Float(if float_item.is_nan() { 0.0 } else { float_item })
+            }
+            4 => Ipld::String(String::arbitrary(g)),
+            5 => Ipld::Bytes(Vec::arbitrary(g)),
+            6 => Ipld::List(Self::arbitrary_ipld_vec(g, size)),
+            7 => Ipld::Map(Self::arbitrary_ipld_map(g, size)),
+            8 => {
+                let cid = Cid::new_v1(
+                    u64::arbitrary(g),
+                    cid::multihash::Multihash::wrap(u64::arbitrary(g), &[u8::arbitrary(g)])
+                        .unwrap(),
+                );
+                Ipld::Link(cid)
+            }
+            // unreachable due to the fact that
+            // we know that the index is always < 9
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_ipld_vec(g: &mut quickcheck::Gen, rec_size: &mut usize) -> Vec<Ipld> {
+        if *rec_size == 0 {
+            return vec![];
+        }
+        *rec_size -= 1;
+        (0..Self::arbitrary_size(g))
+            .map(|_| Self::arbitrary_ipld(g, rec_size))
+            .collect()
+    }
+
+    fn arbitrary_ipld_map(g: &mut quickcheck::Gen, rec_size: &mut usize) -> BTreeMap<String, Ipld> {
+        if *rec_size == 0 {
+            return BTreeMap::new();
+        }
+        *rec_size -= 1;
+        (0..Self::arbitrary_size(g))
+            .map(|_| (String::arbitrary(g), Self::arbitrary_ipld(g, rec_size)))
+            .collect()
+    }
+
+    fn arbitrary_size(g: &mut quickcheck::Gen) -> usize {
+        let s = g.size();
+        let choose_vec = Vec::from_iter(0..s);
+        *g.choose(choose_vec.as_slice()).unwrap()
```

Comment:
You might consider doing something like `usize::arbitrary(g) % g.size()`.

Reply:
Thanks! Fixed